kernels/portable/cpu/util/test/broadcast_indexes_range_test.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/kernels/portable/cpu/util/broadcast_indexes_range.h>
+#include <executorch/kernels/portable/cpu/util/broadcast_util.h>
+#include <executorch/runtime/core/exec_aten/testing_util/tensor_factory.h>
+
+#include <gtest/gtest.h>
+
+using executorch::aten::ScalarType;
+using executorch::aten::Tensor;
+using executorch::runtime::testing::TensorFactory;
+using torch::executor::BroadcastIndexesRange;
+using torch::executor::delinearize_index;
+using torch::executor::linearize_access_indexes;
+
+namespace {
+template <typename Range>
+auto range_to_vec(const Range& rng) {
+  return std::vector<typename Range::iterator::value_type>(
+      rng.begin(), rng.end());
+}
+} // namespace
+TEST(BroadcastIndexesRangeTest, Empty) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor a = tf.make({0}, {});
+  ASSERT_EQ(a.numel(), 0);
+  bool loop_entered = false;
+  for (auto _ : BroadcastIndexesRange<1>(a, a)) {
+    loop_entered = true;
+  }
+  EXPECT_FALSE(loop_entered);
+}
+
+// [W] -> [W]
+TEST(BroadcastIndexesRangeTest, OneDNotBroadcasted) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor out = tf.zeros({5});
+  int idx = 0;
+  for (const auto& elem : range_to_vec(BroadcastIndexesRange<1>(out, out))) {
+    EXPECT_EQ(elem[0], idx++);
+    EXPECT_EQ(elem[0], elem[1]);
+  }
+}
+
+// [1] -> [W]
+TEST(BroadcastIndexesRangeTest, ScalarBroadcastToOneD) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor out = tf.zeros({5});
+  Tensor in = tf.zeros({1});
+
+  auto actual = range_to_vec(BroadcastIndexesRange<1>(out, in));
+  decltype(actual) expected = {
+      {0, 0},
+      {1, 0},
+      {2, 0},
+      {3, 0},
+      {4, 0},
+  };
+  EXPECT_EQ(expected, actual);
+}
+
+// [1] -> [H, W]
+// [1, W] -> [H, W]
+// [H, 1] -> [H, W]
+// [H, W] -> [H, W]
+// Cover all these at the same time to also exercise multiple input tensors.
```

Comment:
you are also covering [1, 1] -> [H, W] and [W] -> [H, W] here. would be good to mention as well.

Reply:
good catch, I'll rename it to OneAndTwoDExhaustive